spec.html

```diff
@@ -7680,6 +7680,99 @@ <h1>Jobs and Host Operations to Enqueue Jobs</h1>
+
+    <emu-clause id="sec-hostgetjobcallbackambientcontext" aoid="HostGetJobCallbackAmbientContext">
+      <h1>HostGetJobCallbackAmbientContext ( )</h1>
+      <p>HostGetJobCallbackAmbientContext is a host-defined abstract operation that allows hosts to provide an ambient context to associate with a callback. This is called at the time that the callback is passed to the function that is responsible for its being eventually scheduled and run.</p>
+      <p>The implementation of HostGetJobCallbackAmbientContext must conform to the following requirements:</p>
+      <ul>
+        <li>It must always complete normally (i.e., not return an abrupt completion).</li>
+      </ul>
+      <p>The default implementation of HostGetJobCallbackAmbientContext returns *undefined*.</p>
+    </emu-clause>
+
+    <emu-clause id="sec-hostgetjobcallbackambientcontext" aoid="HostSetJobCallbackAmbientContext">
+      <h1>HostSetJobCallbackAmbientContext ( _context_ )</h1>
+      <p>HostSetJobCallbackAmbientContext is a host-defined abstract operation that allows hosts to restore the ambient context associated with a callback when it is invoked during the evaluation of a Job Abstract Closure.</p>
+      <p>The implementation of HostSetJobCallbackAmbientContext must conform to the following requirements:</p>
+      <ul>
+        <li>It must always complete normally (i.e., not return an abrupt completion).</li>
+      </ul>
+      <p>The default implementation of HostSetJobCallbackAmbientContext does nothing.</p>
+    </emu-clause>
+
+    <emu-clause id="sec-makejobcallback" aoid="MakeJobCallback">
+      <h1>MakeJobCallback ( _callable_ )</h1>
+      <p>The abstract operation MakeJobCallback takes argument _callable_ (a callable object). It performs the following steps when called:</p>
+      <emu-alg>
+        1. Assert: IsCallable(_callable_) is *true*.
+        1. Return the JobCallback Record { [[Callback]]: _callable_, [[HostAmbientContext]]: ! HostGetJobCallbackAmbientContext() }.
+      </emu-alg>
+      <emu-note>
+        <p>This is called at the time that the callback is passed to the function that is responsible for its being eventually scheduled and run. For example, `Promise.prototype.then(thenAction)` calls MakeJobCallback on `thenAction` at the time of invoking `Promise.prototype.then`, not at the time of scheduling the reaction Job.</p>
+      </emu-note>
+    </emu-clause>
+
+    <emu-clause id="sec-calljobcallback" aoid="CallJobCallback">
+      <h1>CallJobCallback ( _jobCallback_, _V_, [ , _argumentsList_ ] )</h1>
+      <p>The abstract operation CallJobCallback takes argument _jobCallback_ (a JobCallback Record), _V_ (an ECMAScript language value) and optional argument _argumentsList_ (a List of ECMAScript language values). It performs the following steps when called:</p>
+      <emu-alg>
+        1. Assert: IsCallable(_jobCallback_.[[Callback]]) is *true*.
+        1. Perform ! HostSetJobCallbackAmbientContext(_jobCallback_.[[HostAmbientContext]]).
```

Comment:
Assuming HostSetJobCallbackAmbientContext is largely equivalent to [prepare to run a callback](https://html.spec.whatwg.org/multipage/webappapis.html#prepare-to-run-a-callback), Web IDL also has a sibling [clean up after running a callback](https://html.spec.whatwg.org/multipage/webappapis.html#clean-up-after-running-a-callback). Do we need a hook here too?

Reply:
Ah, good point. Looks like we need to both a pre- and a post- hook for invoking job callbacks.

Reply:
Reworked.